tempodb/backend/s3/s3.go

```diff
@@ -32,24 +32,51 @@ type readerWriter struct {
 	core   *minio.Core
 }
 
+type overrideSignatureVersion struct {
+	useV2    bool
+	upstream credentials.Provider
+}
+
+func (s *overrideSignatureVersion) Retrieve() (credentials.Value, error) {
+	v, err := s.upstream.Retrieve()
+	if err != nil {
+		return v, err
+	}
+
+	if s.useV2 && !v.SignerType.IsAnonymous() {
+		v.SignerType = credentials.SignatureV2
+	}
+
+	return v, nil
+}
+
+func (s *overrideSignatureVersion) IsExpired() bool {
+	return s.upstream.IsExpired()
+}
+
 func New(cfg *Config) (backend.Reader, backend.Writer, backend.Compactor, error) {
 	l := log_util.Logger
+
+	wrapCredentialsProvider := func(p credentials.Provider) credentials.Provider {
+		return &overrideSignatureVersion{useV2: cfg.SignatureV2, upstream: p}
```

Comment:
This always wraps the credential Provider.  Thoughts on only wrapping it when SignatureV2 is enabled?  If not enabled then return p with no wrapper.

Reply:
:+1: very good point, latest push contains this change